[Hexagon] Optimize some shuffles

dsharletg opened the pull request:
This PR adds optimizations to certain shuffle_vectors calls (stride 2 ramps, dense ramp selection of native vectors). This moves some old logic from slice_vectors/concat_vectors to avoid needing to handle it in two places (in two different ways).

This does reduce the simulated cycle count of camera_pipe slightly. It also enables a number of tests that were previously failing to pass.

pranavb-ca commented on src/CodeGen_Hexagon.cpp:
```diff
@@ -501,22 +501,27 @@ llvm::Function *CodeGen_Hexagon::define_hvx_intrinsic(llvm::Function *intrin, Ty
 
+Value *CodeGen_Hexagon::create_bitcast(Value *v, llvm::Type *ty) {
+    if (BitCastInst *c = dyn_cast<BitCastInst>(v)) {
+        return create_bitcast(c->getOperand(0), ty);
```
This is strange. Isn't LLVM able to cancel out redundant bitcasts? Is this leading to sub-optimal code?

dsharletg replied:
I'm sure LLVM does this too, the motivation for this isn't to help out LLVM
generate better code from otherwise identical IR, it's to make it so we can
find vcombine and other instructions without having to peel away more than
one bitcast. That lets us generate better IR that LLVM currently can't do
itself.

On Sun, Apr 17, 2016 at 7:01 PM Pranav Bhandarkar <notifications@github.com>
wrote:

> In src/CodeGen_Hexagon.cpp
> <https://github.com/halide/Halide/pull/1129#discussion_r59999344>:
>
> > @@ -501,22 +501,27 @@ llvm::Function *CodeGen_Hexagon::define_hvx_intrinsic(llvm::Function *intrin, Ty
> >      return wrapper;
> >  }
> >
> > +Value *CodeGen_Hexagon::create_bitcast(Value *v, llvm::Type *ty) {
> > +    if (BitCastInst *c = dyn_cast<BitCastInst>(v)) {
> > +        return create_bitcast(c->getOperand(0), ty);
>
> This is strange. Isn't LLVM able to cancel out redundant bitcasts? Is this
> leading to sub-optimal code?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/1129/files/839bf5f0bdcab37fb5459bedd0d685c6b314e293#r59999344>
>

dsharletg commented:
I merged master with this, which includes the interleave_vectors refactoring. @pranavb-ca what do you want to do with this PR relative to #1125 ? I think we should work on #1125, and once that goes into hexagon2, I'll merge hexagon2 into this branch, then we can consider merging this PR.

pranavb-ca commented:
@dsharletg : Your plan sounds good.

dsharletg commented:
I haven't forgotten about this, the merge is proving difficult. I'm working on it.

dsharletg commented:
I pushed the merge, but don't merge yet. The new tests that I had added in this branch are now broken after the aligning loads pass runs. I'm working on it.

dsharletg commented:
I think this is now ready to be looked at. @pranavb-ca, I hope you don't mind, in the course of my debugging of the valign issue, I refactored the aligning loads pretty heavily:

- Changed the logic to early out of the function, instead of a deeply nested if. I think this makes it easier to understand.
- Changed the strided load case to align the dense load if possible. This makes it so it can handle loads like f(2*x + 1) by generating 2 aligned loads and a vpacko, instead of 3 aligned loads, 2 valigns, and a vpacke. This was the start of all of my problems in this merge :) There was a new test in this branch for vpacko for a load like this, that stopped generating when I merged hexagon2 with the aligning loads pass.
- Changed some of the non-native width handles to be "more" recursive, e.g. the unaligned native case now just loads a double wide vector and lets that non-native width handler deal with the concatenation.

dsharletg commented:
OK, I removed the workaround and used the right intrinsic for offsets not fitting in 3 bits. This is ready to be reviewed.

dsharletg commented:
I just merged master, which has the align_storage directive, and then added align_storage directives to the camera pipe. It's all working very nicely! The generated code looks good, and the number of cycles used to run the camera pipe is reduced a bit. Loop store forwarding, along with modifying the schedule to be better for Hexagon, will be a bigger improvement.

dsharletg commented:
The number of cycles to run the camera pipe dropped from 8.04e9 to 7.85e9 with this change.

pranavb-ca commented on test/correctness/simd_op_check.cpp:
```diff
@@ -1327,6 +1327,13 @@ void check_hvx_all() {
+    check("vlalign(v*,v*,#7)", hvx_width/1, in_u8(x + hvx_width - 7));
+    check("valign(v*,v*,r*)", hvx_width/1, in_u8(x + 8));
+    check("valign(v*,v*,r*)", hvx_width/1, in_u8(x + hvx_width - 8));
+
```
It might be useful to test for element_size > 8 bits. This is because the scalar arg of valign is in bytes. So if we are off by 2 elements of 16 bits, then that arg should be 4. This will test if we are moving from elements to bytes correctly when we generate valigns from shuffle_vectors.

dsharletg commented:
Addressed the comment. I also fixed a bug that I had missed until now that was affecting the camera pipe.

pranavb-ca commented on src/CodeGen_Hexagon.cpp:
```diff
@@ -535,92 +540,202 @@ Value *CodeGen_Hexagon::call_intrin_cast(llvm::Type *ret_ty,
+            b_elements = b_ty->getVectorNumElements();
+        }
+        if (start == 0 && result_ty == a_ty) {
+            return a;
```
Consider a shuffle of the type 
shuffle_vectors(dbl_vector_a_b, 0, 31, 1, 32, 2, 33..., 31,63)  
It is a weird interleaving operation that is just a contrived example that won't be equivalent to returning a. If you think this is far too contrived a case to appear in the real world, then please ignore this comment.

dsharletg replied:
I think we'd expect something like that to be coming from interleave_vectors, which is still handled in a separate code path (CodeGen_Hexagon::interleave_vectors), it doesn't go through shuffle_vectors.

The general point stands though, there's quite a lot of shuffle_vectors calls that we could map to HVX instructions, but we currently do not. See the todo on line 735...

pranavb-ca commented on src/AlignLoads.cpp:
```diff
@@ -92,8 +92,7 @@ class AlignLoads : public IRMutator {
             // Load a dense vector covering all of the addresses in the load.
-            internal_assert(stride >= 0);
-            Expr dense_base = simplify(ramp->base - aligned_offset);
+            Expr dense_base = simplify(ramp->base - shift);
```
Thanks for fixing this. I had noticed the inconsistency in the use of aligned_offset here and shift below in the making of the index. I thought it was suspicious, but some how convinced myself that it was ok. I guess I was wrong :)

pranavb-ca commented:
@dsharletg : Aside from that one comment about shuffle_vectors that arises out of the use of max_element in that contrived example, this LGTM.
